assembly/src/tokens/stream.rs

```diff
@@ -25,24 +26,44 @@ impl<'a> TokenStream<'a> {
             return Err(ParsingError::empty_source());
         }
         let mut tokens = Vec::new();
-        let mut doc_comments = BTreeMap::new();
+        let mut proc_comments = BTreeMap::new();
+        let mut module_comment = None;
 
-        let mut comment = Comment(None);
+        let mut comment_builder = CommentBuilder(None);
 
         for line in source.lines() {
             let line = line.trim();
             if line.starts_with(DOC_COMMENT_PREFIX) {
-                comment.append_line(line);
-            } else if line.is_empty() && !comment.is_empty() {
-                return Err(ParsingError::malformed_doc_comment(tokens.len()));
-            } else {
-                if !comment.is_empty() {
-                    doc_comments.insert(tokens.len(), comment.take_content());
+                comment_builder.append_line(line);
+            } else if line.starts_with(LINE_COMMENT_PREFIX) {
+                continue;
+            } else if line.is_empty() {
+                if !comment_builder.is_empty() {
+                    if tokens.is_empty() && module_comment.is_none() {
+                        // if we haven't read any tokens yet, but already have built a comment, a
+                        // new line must indicate the end of a module comment.
+                        module_comment = comment_builder.take_comment();
+                    } else {
+                        // if procedure comment is not immediately followed by anything.
+                        return Err(ParsingError::dangling_procedure_comment(tokens.len()));
+                    }
                 }
+            } else {
                 let mut line_tokens = line
                     .split_whitespace()
                     .take_while(|&token| !token.starts_with(LINE_COMMENT_PREFIX))
                     .collect::<Vec<_>>();
+
+                if !comment_builder.is_empty() {
+                    // procedure comment should always be followed by a procedure token
+                    if line_tokens[0].starts_with(Token::EXPORT)
+                        || line_tokens[0].starts_with(Token::PROC)
```

Comment:
minor nit: i would probably declare a variable for the first token here to make this a bit more readable. Something like:
```Rust
let token = line_tokens[0];
if token.starts_with(Token::EXPORT) || token.starts_with(Token::PROC) {
```